File: cpp/easy/diameter_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root)
            return 0;
        int ret = INT_MIN;
        get_height(root, ret);
        return ret;
    }
    
    int get_height(TreeNode* root, int& ret)
    {
        if(!root)
            return 0;
        int left = get_height(root->left, ret);
        int right = get_height(root->right, ret);
        ret = max(ret, left+right);
        return max(left, right)+1;
    }
};


Conclusion:
caculate the sum of each node's left subtree depth and right subtree depth, get the maxim value.

Time complexity:
O(n) each node will be visit only one time through the function "get_height()"

File: python/easy/diameter_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import sys
class Solution(object):
    ret = -sys.maxint-1
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(root):
            if not root:
                return 0
            left_h = dfs(root.left)
            right_h = dfs(root.right)
            self.ret = max(self.ret, left_h+right_h)
            return max(left_h, right_h)+1
        
        if not root:
            return 0
        dfs(root)
        return self.ret
